protocols/autonat/src/behaviour.rs

```diff
@@ -0,0 +1,853 @@
+                peer_id.unwrap(),
+                error
+            );
+
+            let response = DialResponse {
+                result: Err(ResponseError::DialError),
+                status_text: Some("dial failed".to_string()),
+            };
+            let _ = self.inner.send_response(channel, response);
+        }
+    }
+
+    fn inject_disconnected(&mut self, peer: &PeerId) {
+        self.inner.inject_disconnected(peer);
+        self.connected.remove(peer);
+    }
+
+    fn inject_address_change(
+        &mut self,
+        peer: &PeerId,
+        conn: &ConnectionId,
+        old: &ConnectedPoint,
+        new: &ConnectedPoint,
+    ) {
+        self.inner.inject_address_change(peer, conn, old, new);
+
+        self.connected
+            .insert(*peer, new.get_remote_address().clone());
+    }
+
+    fn inject_new_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_new_listen_addr(id, addr);
+
+        // New address could be publicly reachable, trigger retry.
+        if !self.nat_status.is_public() {
+            if self.confidence > 0 {
+                self.confidence -= 1;
+            }
+            self.schedule_next_probe(self.config.retry_interval);
+        }
+    }
+
+    fn inject_expired_listen_addr(&mut self, id: ListenerId, addr: &Multiaddr) {
+        self.inner.inject_expired_listen_addr(id, addr);
+        if let Some(public_address) = self.public_address() {
+            if public_address == addr {
+                self.confidence = 0;
+                self.nat_status = NatStatus::Unknown;
+                self.schedule_probe.reset(Duration::ZERO);
+            }
+        }
+    }
+
+    fn inject_new_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_new_external_addr(addr);
+
+        // New address could be publicly reachable, trigger retry.
+        if !self.nat_status.is_public() {
+            if self.confidence > 0 {
+                self.confidence -= 1;
+            }
+            self.schedule_next_probe(self.config.retry_interval);
+        }
+    }
+
+    fn inject_expired_external_addr(&mut self, addr: &Multiaddr) {
+        self.inner.inject_expired_external_addr(addr);
+        if let Some(public_address) = self.public_address() {
+            if public_address == addr {
+                self.confidence = 0;
+                self.nat_status = NatStatus::Unknown;
+                self.schedule_probe.reset(Duration::ZERO);
+            }
+        }
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        params: &mut impl PollParameters,
+    ) -> Poll<NetworkBehaviourAction<Self::OutEvent, Self::ProtocolsHandler>> {
+        let mut is_probe_ready = false;
+        loop {
+            if self.schedule_probe.poll_unpin(cx).is_ready() {
+                is_probe_ready = true;
+                self.schedule_probe.reset(self.config.retry_interval);
+                continue;
+            }
```

Comment:
In this concrete case (in `NetworkBehaviour::poll`) we need to make sure to never return `Poll::Pending` without previously having registered the current `Waker` with the `Delay` in `self.schedule_probe`. I am guessing that this is the intention behind `is_probe_ready` and the `continue`, correct?

What do you think of https://github.com/libp2p/rust-libp2p/commit/cb797abd96e2032b98a9f5307215bb8be18d3cd7 instead? Would that solve the same issue? I would find it easier to understand.

Reply:
Polling the waker after polling the inner behaviour is a good idea. However, depending on the request/ responses that are polled from the inner behaviour, one or multiple `Event`s may have been added to `Behaviour::pending_out_events`, therefore this queue also has to be checked before returning `Poll::Pending`.
I refactored the poll function a bit in dbfc59e: 
1. rearranged it according to you proposal so that the waker is polled after the inner behaviour
2. make it more consistent that all `pending_out_events.push_back(..)` happen within the poll method. Before it was a bit confusing when/ where new events are added to the queue since it was spread out onto different methods.

However, I am not really happy with that solution yet; will try to clean it a bit more in the next days.